File: ocmsdk/src/main/java/com/gigigo/orchextra/core/data/api/mappers/elementcache/ApiElementCacheMapper.java
package com.gigigo.orchextra.core.data.api.mappers.elementcache;


import com.gigigo.orchextra.core.data.api.mappers.elements.ApiElementSegmentationMapper;
import com.gigigo.orchextra.core.domain.entities.elementcache.ElementCacheType;
import com.gigigo.orchextra.core.data.api.dto.elementcache.ApiElementCache;
import com.gigigo.orchextra.core.domain.entities.elementcache.ElementCache;
import com.gigigo.ggglib.mappers.ExternalClassToModelMapper;
import java.util.ArrayList;
import java.util.List;

public class ApiElementCacheMapper
    implements ExternalClassToModelMapper<ApiElementCache, ElementCache> {

  private final ApiElementCacheRenderMapper apiElementCacheItemRenderMapper;
  private final ApiElementCachePreviewMapper apiElementCachePreviewMapper;
  private final ApiElementCacheShareMapper apiElementCacheShareMapper;
  private final ApiElementSegmentationMapper apiElementSegmentationMapper;

  public ApiElementCacheMapper(ApiElementCacheRenderMapper apiElementCacheItemRenderMapper,
      ApiElementCachePreviewMapper apiElementCachePreviewMapper,
      ApiElementCacheShareMapper apiElementCacheShareMapper,
      ApiElementSegmentationMapper apiElementSegmentationMapper) {
    this.apiElementCacheItemRenderMapper = apiElementCacheItemRenderMapper;
    this.apiElementCachePreviewMapper = apiElementCachePreviewMapper;
    this.apiElementCacheShareMapper = apiElementCacheShareMapper;
    this.apiElementSegmentationMapper = apiElementSegmentationMapper;
  }

  @Override public ElementCache externalClassToModel(ApiElementCache data) {
    ElementCache model = new ElementCache();

    model.setSlug(data.getSlug());
    model.setType(ElementCacheType.convertStringToEnum(data.getType()));
    model.setName(data.getName());
    model.setUpdateAt(data.getUpdatedAt());

    List<String> tagList = new ArrayList<>();
    if (data.getTags() != null) {
      for (String tag : data.getTags()) {
        tagList.add(tag);
      }
      model.setTags(tagList);
    }

    if (data.getRender() != null) {
      model.setRender(apiElementCacheItemRenderMapper.externalClassToModel(data.getRender()));
    }

    if (data.getPreview() != null) {
      model.setPreview(apiElementCachePreviewMapper.externalClassToModel(data.getPreview()));
    }

    if (data.getShare() != null) {
      model.setShare(apiElementCacheShareMapper.externalClassToModel(data.getShare()));
    }

    if (data.getSegmentation() != null) {
      model.setSegmentation(apiElementSegmentationMapper.externalClassToModel(data.getSegmentation()));
    }

    return model;
  }
}

File: ocmsdk/src/main/java/com/gigigo/orchextra/core/data/api/mappers/elementcache/ApiElementCacheMapper.kt
package com.gigigo.orchextra.core.data.api.mappers.elementcache

import android.util.Log
import com.gigigo.ggglib.mappers.ExternalClassToModelMapper
import com.gigigo.orchextra.core.data.api.dto.elementcache.ApiElementCache
import com.gigigo.orchextra.core.domain.entities.elementcache.ElementCache
import com.gigigo.orchextra.core.domain.entities.elementcache.ElementCacheType
import java.util.ArrayList

class ApiElementCacheMapper(
    private val apiElementCacheItemRenderMapper: ApiElementCacheRenderMapper,
    private val apiElementCachePreviewMapper: ApiElementCachePreviewMapper,
    private val apiElementCacheShareMapper: ApiElementCacheShareMapper) : ExternalClassToModelMapper<ApiElementCache, ElementCache> {

  override fun externalClassToModel(data: ApiElementCache): ElementCache {
    val time = System.currentTimeMillis()

    val model = ElementCache()

    with(model) {
      slug = data.slug
      type = ElementCacheType.convertStringToEnum(data.type)
      name = data.name
      updateAt = data.updatedAt
    }

    val tagList = ArrayList<String>()

    with(data) {
      tags?.let {
        for (tag in tags) {
          tagList.add(tag)
        }
        model.tags = tagList
      }

      render?.let {
        model.render = apiElementCacheItemRenderMapper.externalClassToModel(render)
      }

      preview?.let {
        model.preview = apiElementCachePreviewMapper.externalClassToModel(preview)
      }

      share?.let {
        model.share = apiElementCacheShareMapper.externalClassToModel(share)
      }

      customProperties?.let {
        model.customProperties = data.customProperties
      }
    }

    val currentTime = System.currentTimeMillis() - time
    Log.v("TT - ApiElementCache", ("" + currentTime / 1000))

    return model
  }
}
